Updated MLP User Guide (#7059)
mlflow/mlflow: docs/source/pipelines.rst

      pip install mlflow[pipelines]

>   Note that MLflow Pipelines requires [Make](https://www.gnu.org/software/make/),

Reviewer: Think this needs to be formatted with 
````
`Make <https://www.gnu.org/software/make>`
````

  which may not be preinstalled on some Windows systems.
  Windows users must install Make before using MLflow Pipelines. For more information about

>   installing Make on Windows, see https://gnuwin32.sourceforge.net/install.html.

Reviewer 2: While we're here, can we add an extra newline between the next sentence (line 47) and the databricks section (line 49)? Didn't realize, but the text is mashed together a bit:

![Uploading Screen Shot 2022-10-20 at 5.49.29 PM.png…]()

Reviewer 2: Got it. Can you try https://stackoverflow.com/a/46528727/11952869?

Author: Adding a `|` in a new line doesn't work: 
`Bullet list ends without a blank line; unexpected unindent.`
Or simply adding a real `|` to the end of the first bullet.

Reviewer: Dumb question but a newline is two spaces in markdown, is it the same here?

Author: Tried. No effect.

Reviewer 2: Can you look into this further and resolve before merging? Perhaps the solutions in https://stackoverflow.com/questions/54001054/force-a-new-line-in-sphinx-text will work.

Usage
-----
Development environments
~~~~~~~~~~~~~~~~~~~~~~~~

> We recommend to develop models with MLflow Pipelines with one of the following environment settings:

Reviewer 2: ```suggestion
We recommend using one of the following environment configurations to develop models with MLflow Pipelines:
```

[**Databricks**]
  - Edit YAML config and Python files in Databricks Repos. Open separate browser tabs for each

>     file module that you wish to modify.

Reviewer 2: ```suggestion
    file that you wish to modify.
```

>   - Use ``notebooks/databricks.py`` as the driver to run pipeline steps and inspect its output.

Reviewer 3: It feels like these instructions are missing a few steps for project construction and git repo initialization. Should we first tell users to fork the template repo and/or pull the repo into Databricks Repos? Otherwise, we might give off the impression that we expect users to construct the YAML, python, and notebook files on their own. This also applies to the other environment settings.

Reviewer 3: Oh, I see that cloning the template is mentioned below in the `Model development workflow` section. Maybe that section should be above the `Development environments` section?

Reviewer 2: Agreed, we should start with how to clone the repo on Databricks and then talk about which tabs to open.

Reviewer 2: Shall we also tell users to open the notebook in a separate tab and also explain that this `databricks.py` file is converted to a notebook when the repo is cloned with Databricks?

Author: Good point! Swapped the two sections. On explaining "explaining that this databricks.py file is converted to a notebook when the repo is cloned with Databricks", I'm not sure how to best phrase it. In addition, is it a useful information to let users know about this product detail? As long as they open `notebooks/databricks.py`, they should realize what's going on.

  - Pin the workspace browser for easy file navigation.

>   .. image:: _static/images/pipelines_databricks_ui.png

Reviewer 2: Very nitty, but can we edit the screenshot so that it doesn't show "E2 Dogfood" with a doge icon?

Author: Will change it to DB logo.

Author: Done.

    :width: 60%

[**Local with Jupyter Notebook**]

>   - Use ``notebooks/jupyter.ipynb`` as the driver to run pipeline steps and inspect its output.

Reviewer 2: ```suggestion
  - Use ``notebooks/jupyter.ipynb`` as the driver to run pipeline steps and inspect their outputs.
```

  - Use commandline such as ``mlflow pipelines run --profile local`` to run the entire pipeline

>     after done with model quality iteration.

Reviewer 3: Where would I run this command? At the project root?

Reviewer 3: Hmmm, why would a user want to run the entire pipeline once after model iteration? Is it to see if the pipeline can be run as a job?

Reviewer 2: Not sure what this means:

```
after done with model quality iteration.
```

Can we clarify?

Author: Updated. I'm trying to provide alternative ways to run the pipeline (instead of through the notebook).

[**Edit locally with IDE (VSCode) and run on Databricks**]
  - Edit files on your local machine with VSCode and Jupyter plugin.
  - Use |dbx| to sync them to |Databricks Repos| as demonstrated below.

>   - On Databricks, use the ``notebooks/databricks.py`` as the driver to run pipeline steps and inspect its output.

Reviewer 2: ```suggestion
  - On Databricks, use the ``notebooks/databricks.py`` notebook as the driver to run pipeline steps and inspect their outputs.
```

   creates and displays an interactive **Step Card** with the results of the last executed
   :ref:`step <steps-key-concept>`.
   Each :ref:`Pipeline Template <pipeline-templates>` also includes a |Databricks Notebook|
   and a |Jupyter Notebook| for running the pipeline and inspecting its results.

    .. figure:: _static/images/pipelines_evaluate_step_card.png
      :width: 60%

      An example step card produced by running the **evaluate** step of the
      :ref:`MLflow Regression Pipeline <mlflow-regression-pipeline>`. The step card results
      indicate that the trained model passed all performance validations and is ready for
      registration with the :ref:`MLflow Model Registry <registry>`.

    .. figure:: _static/images/pipelines_databricks_logged_artifacts.png
      :width: 60%

      An example MLflow run view page, showing that artifacts from pipeline steps are mostly logged
      and sharable through the MLflow experiment.
 
    .. note::
      Data profiling is often best viewed with "quantiles" mode. To switch it on, on the Facet
      data profile, find ``Chart to show``, click the selector below, and choose ``Quantiles``.

    |

>     .. code-section::

Reviewer 2: Can we move the code section to the top of step 3 so that the results are shown after the MLP execution code? I think most users are probably looking for info about what code to run first.

Author: Done. Also moved NOTE to the last part of the step 3.

Usage
-----
Model development workflow
~~~~~~~~~~~~~~~~~~~~~~~~~~

The general model development workflow for MLflow Pipelines can be summarized through the following

> 4 steps:

Reviewer: I think the prior text worked fine, albeit with "model development":
`The general model development workflow for using MLflow Pipelines is as follows:`

1. Clone a :ref:`Pipeline Template <pipeline-templates>` git repository corresponding to the ML
   problem that you wish to solve. Follow the template's README file for template-specific
   instructions.

>     - [Local] See example below for cloning the |MLflow Pipelines Regression Template| on a local
>       machine.

Reviewer: ```suggestion
    - [Local] Clone the |MLflow Pipelines Regression Template| into a local directory.```

    .. code-block:: sh

      git clone https://github.com/mlflow/mlp-regression-template

    - [Databricks] Clone the |MLflow Pipelines Regression Template| git repository using |Databricks Repos|.

      .. image:: _static/images/pipelines_databricks_repo_ui.png
        :width: 60%

> 2. Edit required fields by searching comments ``FIXME::REQUIRED`` in ``pipeline.yaml`` and

Reviewer: ```suggestion
2. Edit required fields marked by ``FIXME::REQUIRED`` comments in ``pipeline.yaml`` and
```

          # Inspect the resulting model performance evaluations
          mlflow pipelines inspect --step evaluate --profile local

    .. figure:: _static/images/pipelines_evaluate_step_card.png
      :width: 60%

      An example step card produced by running the **evaluate** step of the
      :ref:`MLflow Regression Pipeline <mlflow-regression-pipeline>`. The step card results
      indicate that the trained model passed all performance validations and is ready for
      registration with the :ref:`MLflow Model Registry <registry>`.

    .. figure:: _static/images/pipelines_databricks_logged_artifacts.png
      :width: 60%

      An example MLflow run view page, showing artifacts logged from the Pipeline's steps.
 
    .. figure:: _static/images/pipelines_databricks_notebook_ui.png
      :scale: 25

      Example pipeline run from the |Databricks Notebook| included in the
      |MLflow Pipelines Regression Template|.

    .. note::
      Data profiling is often best viewed with "quantiles" mode. To switch it on, on the Facet
      data profile, find ``Chart to show``, click the selector below, and choose ``Quantiles``.

> 4. Iterate over step 2 and 3: make changes in an individual step, and test them by running

Reviewer 2: ```suggestion
4. Iterate over step 2 and 3: make changes to an individual step, and test them by running
```

          # Inspect the resulting model performance evaluations
          mlflow pipelines inspect --step evaluate --profile local

    .. figure:: _static/images/pipelines_evaluate_step_card.png
      :width: 60%

      An example step card produced by running the **evaluate** step of the
      :ref:`MLflow Regression Pipeline <mlflow-regression-pipeline>`. The step card results

>       indicate that the trained model passed all performance validations and is ready for

Reviewer: Not a comment, but github thought `all` was important and I thought it looked funny
<img width="338" alt="image" src="https://user-images.githubusercontent.com/51172624/197676902-f4791249-e76a-4e9b-9090-d11986868929.png">

      registration with the :ref:`MLflow Model Registry <registry>`.

    .. figure:: _static/images/pipelines_databricks_logged_artifacts.png
      :width: 60%

      An example MLflow run view page, showing that artifacts from pipeline steps are mostly logged

>       and sharable through the MLflow experiment.

Reviewer: Maybe just `showing artifacts logged from the Pipeline's steps`?

      local filesystem. The ``MLFLOW_PIPELINES_EXECUTION_DIRECTORY`` environment variable can
      be used to specify an alternative location for caching results.

Development environments
~~~~~~~~~~~~~~~~~~~~~~~~
We recommend using one of the following environment configurations to develop models with MLflow Pipelines:

[**Databricks**]
  - Edit YAML config and Python files in Databricks Repos. Open separate browser tabs for each
    file module that you wish to modify. For example,
    one for the pipeline config file ``pipeline.yaml``,
    one for the profile config file ``profile/databricks.yaml``,
    one for the driver notebook ``notebooks/databricks.py``,
    and one for the current step (e.g. train) under development ``steps/train.py``.
  - Use ``notebooks/databricks.py`` as the driver to run pipeline steps and inspect its output.
  - Pin the workspace browser for easy file navigation.

  .. image:: _static/images/pipelines_databricks_ui.png
    :width: 60%

[**Local with Jupyter Notebook**]
  - Use ``notebooks/jupyter.ipynb`` as the driver to run pipeline steps and inspect its output.

>   - Edit ``pipeline.yaml``, ``steps/*.py`` and ``profiles/*.yaml`` accordingly with an IDE of your
>     choice.

Reviewer 2: ```suggestion
  - Edit ``pipeline.yaml``, ``steps/*.py`` and ``profiles/*.yaml`` accordingly with an editor of your
    choice.
```

>   - To run the entire pipeline, either user ``notebooks/jupyter.ipynb`` or on commandline with
>     ``mlflow pipelines run --profile local`` (change root to the project root first).

Reviewer 2: ```suggestion
  - To run the entire pipeline, either run ``notebooks/jupyter.ipynb`` or, on the commandline, invoke
    ``mlflow pipelines run --profile local`` (change the current working directory to the project root first).
```

Author: Added a few sentences in step 3 to indicate the use of profiles.

          # Inspect the resulting model performance evaluations
          mlflow pipelines inspect --step evaluate --profile local

      .. figure:: _static/images/pipelines_databricks_notebook_ui.png
        :scale: 25

        Example pipeline run from the |Databricks Notebook| included in the
        |MLflow Pipelines Regression Template|.

4. Iterate over step 2 and 3: make changes in an individual step, and test them by running
   the step and observing the results it produces.
   Use ``Pipeline.inspect()`` to visualize the overall Pipeline dependency graph and artifacts
   each step produces.

>    Use ``Pipeline.get_artifact()`` API to further inspect the step output in a notebook.

Reviewer 2: ```suggestion
   Use ``Pipeline.get_artifact()`` to further inspect individual step outputs in a notebook.
```

What about in a script or CLI?
